File: threedollar-api-user/src/main/java/com/depromeet/threedollar/api/user/config/interceptor/LoginCheckHandler.java
package com.depromeet.threedollar.api.user.config.interceptor;

import com.depromeet.threedollar.api.user.config.session.SessionConstants;
import com.depromeet.threedollar.common.exception.model.UnAuthorizedException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.session.Session;
import org.springframework.session.SessionRepository;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import javax.servlet.http.HttpServletRequest;

@RequiredArgsConstructor
@Component
public class LoginCheckHandler {

    private final SessionRepository<? extends Session> sessionRepository;

    Long getUserId(HttpServletRequest request) {
        String sessionId = request.getHeader(HttpHeaders.AUTHORIZATION);
        if (StringUtils.hasLength(sessionId)) {
            Session session = findSessionBySessionId(sessionId);
            Long userId = session.getAttribute(SessionConstants.USER_ID);
            if (userId != null) {
                return userId;
            }
        }
        throw new UnAuthorizedException(String.format("잘못된 세션 (%S) 입니다", sessionId));
    }

    private Session findSessionBySessionId(String sessionId) {
        Session session = sessionRepository.findById(sessionId);
        if (session == null) {
            throw new UnAuthorizedException(String.format("잘못된 세션 (%S) 입니다", sessionId));
        }
        return session;
    }

}

File: threedollar-api-boss/src/main/kotlin/com/depromeet/threedollar/api/boss/config/interceptor/LoginCheckHandler.kt
package com.depromeet.threedollar.api.boss.config.interceptor

import javax.servlet.http.HttpServletRequest
import org.springframework.http.HttpHeaders
import org.springframework.session.Session
import org.springframework.session.SessionRepository
import org.springframework.stereotype.Component
import com.depromeet.threedollar.api.boss.config.session.SessionConstants
import com.depromeet.threedollar.common.exception.model.ForbiddenException
import com.depromeet.threedollar.common.exception.model.UnAuthorizedException
import com.depromeet.threedollar.common.exception.type.ErrorCode
import com.depromeet.threedollar.domain.mongo.boss.domain.account.BossAccountRepository
import com.depromeet.threedollar.domain.mongo.boss.domain.registration.BossRegistrationRepository

private const val HEADER_BEARER_PREFIX = "Bearer "

@Component
class LoginCheckHandler(
    private val sessionRepository: SessionRepository<out Session?>,
    private val bossAccountRepository: BossAccountRepository,
    private val bossRegistrationRepository: BossRegistrationRepository,
) {

    fun checkAuthOptional(request: HttpServletRequest): Boolean {
        val authorization = request.getHeader(HttpHeaders.AUTHORIZATION) ?: return true
        if (authorization.split(HEADER_BEARER_PREFIX).size <= 1) {
            return true
        }
        val sessionId = sessionRepository.findById(authorization.split(HEADER_BEARER_PREFIX)[1])
        val bossAccountId: String? = sessionId?.getAttribute(SessionConstants.BOSS_ACCOUNT_ID)
        bossAccountId?.let {
            if (bossAccountRepository.existsBossAccountById(bossAccountId)) {
                request.setAttribute(SessionConstants.BOSS_ACCOUNT_ID, bossAccountId)
            }
        }
        return true
    }

    fun checkAuthRequired(request: HttpServletRequest): Boolean {
        val header = request.getHeader(HttpHeaders.AUTHORIZATION)
            ?: throw UnAuthorizedException("인증이 실패하였습니다 - 빈 Authorization 헤더가 요청되었습니다")

        if (!header.startsWith(HEADER_BEARER_PREFIX)) {
            throw UnAuthorizedException("인증이 실패하였습니다 - ($HEADER_BEARER_PREFIX) 형식이 아닌 Authorization 헤더($header)입니다.")
        }

        val sessionId = header.split(HEADER_BEARER_PREFIX)[1]
        val bossAccountId: String = findSessionBySessionId(sessionId).getAttribute(SessionConstants.BOSS_ACCOUNT_ID)

        if (bossAccountRepository.existsBossAccountById(bossAccountId)) {
            request.setAttribute(SessionConstants.BOSS_ACCOUNT_ID, bossAccountId)
            return true
        }
        bossRegistrationRepository.findWaitingRegistrationById(bossAccountId)?.let {
            throw ForbiddenException("현재 가입 승인 대기중인 사장님($bossAccountId) 입니다", ErrorCode.FORBIDDEN_WAITING_APPROVE_BOSS_ACCOUNT)
        } ?: throw UnAuthorizedException("해당하는 사장님 계정($bossAccountId) 혹은 대기중인 가입 신청($bossAccountId)은 존재하지 않습니다")
    }

    private fun findSessionBySessionId(sessionId: String): Session {
        return sessionRepository.findById(sessionId)
            ?: throw UnAuthorizedException("인증이 실패하였습니다 - 해당하는 세션($sessionId)은 존재하지 않습니다.")
    }

}
